skiko/src/awtMain/kotlin/org/jetbrains/skiko/redrawer/MetalRedrawer.kt

```diff
@@ -73,63 +72,82 @@ internal class MetalRedrawer(
         }
         _device = initDevice
         contextHandler = MetalContextHandler(layer, initDevice, adapter)
-        setDisplaySyncEnabled(initDevice.ptr, properties.isVsyncEnabled)
+        setDisplaySyncEnabled(initDevice.ptr, false)  // VSync is managed manually (see vsyncedFrameDispatcher)
     }
 
     override val renderInfo: String get() = contextHandler.rendererInfo()
 
     private val frameDispatcher = FrameDispatcher(MainUIDispatcher) {
         if (layer.isShowing) {
-            update(System.nanoTime())
-            draw()
+            update()
+            drawOnlyFrameDispatcher.scheduleFrame()
         }
     }
 
+    private val throttledFrameDispatcher = FrameDispatcher(MainUIDispatcher) {
+        if (layer.isShowing) {
+            update()
+            drawOnlyFrameDispatcher.scheduleFrame()
```

Comment:
I have concerns that we schedule drawing here, not calling it immediately. We might skip some draw's, making jumpy pacing `update - update - draw - update - update - draw` instead of `update - draw - update - draw`:

```
frameDispatcher1
  frameDispatcher.update
    ...
    frameDispatcher.needRedraw(throttledToVsync= false) // schedules frameDispatcher2
  drawOnlyFrameDispatcher.scheduleFrame() // will be called after frameDispatcher2

frameDispatcher2
drawOnlyFrameDispatcher // lag, to late for draw, we had to draw before frameDispatcher2
frameDispatcher3
...
```

Reply:
If we draw immediately here, the `temporary change is not visible` test fails because a running animation causes the temporary change to be drawn.

I'm not sure I understand your concern, though. This whole PRs purpose is to allow a sequence like `update - update - draw`. It makes sure that drawing only happens on vsync.

Reply:
> temporary change is not visible

Looking at the test, I see issues it is incorrect requirement that we shouldn't draw/show temporary updates. If something is updated, and it requested a draw, we should draw it ASAP, to avoid a large latency between the state update and the actual change on the screen.

in this code:
```
renderDelegate.color = tempColor
renderDelegate.layer.needRedraw(throttledToVsync = true)
renderChannel.receive()  // Wait until render was actually called
renderDelegate.color = color
renderDelegate.layer.needRedraw(throttledToVsync = true)
```
if there was a vsync after `renderChannel.receive()`, we should draw it.

Reply:
> This whole PRs purpose is to allow a sequence like update - update - draw

The is also another change in this PR - not waiting for vsync for some updates. It is good change.

But `update - update - draw` as I said above adds latencies, and can introduce frame drops. Ideally every update should be followed by a draw, so we can see the actual state on the screen (it can be not achievable on macOs because of lost frames issue)

Reply:
> (it can be not achievable on macOs because of lost frames issue

There is a theoretical way (don't sent the last buffer until 2 others are in flight, and reuse it for new content). But this is out of scope of this PR

Reply:
>  to avoid a large latency between the state update and the actual change on the screen.

I remember we discussed online that it is to avoid issues for blinking. But I believe that it should be fixed not by the renderer logic, but by the application logic. Because it is not achievable in the renderer logic, we can only add a latency that just reduce frequency of blinking, but not eliminate fully.